write: broken pipe panic when receiver runs out of disk space

Firstly, thank you for this excellent software! It's fantastically helpful!

I was testing croc to send a file over my LAN and sender panicked when the receiver ran out of disk space:

```
[error] 2024/08/29 22:05:48 tcp.go:445: write error on channel 2: write tcp4 192.168.50.237:9012->192.168.50.101:34770: write: broken pipe
[error] 2024/08/29 22:05:48 tcp.go:445: write error on channel 2: write tcp4 192.168.50.237:9011->192.168.50.101:52906: write: broken pipe
[error] 2024/08/29 22:05:48 tcp.go:445: write error on channel 2: write tcp4 192.168.50.237:9011->192.168.50.101:52906: write: broken pipe
[error] 2024/08/29 22:05:48 tcp.go:445: write error on channel 2: write tcp4 192.168.50.237:9011->192.168.50.101:52906: write: broken pipe
[error] 2024/08/29 22:05:48 tcp.go:445: write error on channel 2: write tcp4 192.168.50.237:9010->192.168.50.101:40064: write: broken pipe
[error] 2024/08/29 22:05:48 tcp.go:445: write error on channel 2: write tcp4 192.168.50.237:9010->192.168.50.101:40064: write: broken pipe
panic: connection.Write failed: write tcp 127.0.0.1:51475->127.0.0.1:9012: write: broken pipe

goroutine 133 [running]:
github.com/schollz/croc/v10/src/croc.(*Client).sendData(0x1400022c008, 0x2)
        github.com/schollz/croc/v10/src/croc/croc.go:2063 +0x800
created by github.com/schollz/croc/v10/src/croc.(*Client).updateState in goroutine 31
        github.com/schollz/croc/v10/src/croc/croc.go:1889 +0x500
```

The sender is running:

```console
$ croc --version
croc version v10.0.11
$ uname -a
Darwin Toms-Laptop.local 23.6.0 Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:04 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T8122 arm64
```

The receiver was:

```console
$ croc --version
croc version v10.0.11
uname -a
Linux ubuntu 6.8.0-41-generic #41-Ubuntu SMP PREEMPT_DYNAMIC Fri Aug  2 20:41:06 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

Let me know if you'd like more information.

--- a/src/croc/croc.go
+++ b/src/croc/croc.go
@@ -31,6 +31,7 @@ import (
 	"github.com/schollz/croc/v10/src/comm"
 	"github.com/schollz/croc/v10/src/compress"
 	"github.com/schollz/croc/v10/src/crypt"
+	"github.com/schollz/croc/v10/src/diskusage"
 	"github.com/schollz/croc/v10/src/message"
 	"github.com/schollz/croc/v10/src/models"
 	"github.com/schollz/croc/v10/src/tcp"
@@ -1268,6 +1269,12 @@ func (c *Client) processMessageFileInfo(m message.Message) (done bool, err error
 			}
 		}
 	}
+	// check the totalSize does not exceed disk space
+	usage := diskusage.NewDiskUsage(".")
+	if usage.Available() < uint64(totalSize) {
+		return true, fmt.Errorf("not enough disk space")
+	}
+
 	// c.spinner.Stop()
 	action := "Accept"
 	if c.Options.SendingText {

--- a/src/diskusage/diskusage.go
+++ b/src/diskusage/diskusage.go
@@ -1,0 +1,49 @@
+//go:build !windows
+// +build !windows
+
+package diskusage
+
+import (
+	"golang.org/x/sys/unix"
+)
+
+// DiskUsage contains usage data and provides user-friendly access methods
+type DiskUsage struct {
+	stat *unix.Statfs_t
+}
+
+// NewDiskUsages returns an object holding the disk usage of volumePath
+// or nil in case of error (invalid path, etc)
+func NewDiskUsage(volumePath string) *DiskUsage {
+	stat := unix.Statfs_t{}
+	err := unix.Statfs(volumePath, &stat)
+	if err != nil {
+		return nil
+	}
+	return &DiskUsage{&stat}
+}
+
+// Free returns total free bytes on file system
+func (du *DiskUsage) Free() uint64 {
+	return du.stat.Bfree * uint64(du.stat.Bsize)
+}
+
+// Available return total available bytes on file system to an unprivileged user
+func (du *DiskUsage) Available() uint64 {
+	return uint64(du.stat.Bavail) * uint64(du.stat.Bsize)
+}
+
+// Size returns total size of the file system
+func (du *DiskUsage) Size() uint64 {
+	return uint64(du.stat.Blocks) * uint64(du.stat.Bsize)
+}
+
+// Used returns total bytes used in file system
+func (du *DiskUsage) Used() uint64 {
+	return du.Size() - du.Free()
+}
+
+// Usage returns percentage of use on the file system
+func (du *DiskUsage) Usage() float32 {
+	return float32(du.Used()) / float32(du.Size())
+}

--- a/src/diskusage/diskusage_test.go
+++ b/src/diskusage/diskusage_test.go
@@ -1,0 +1,17 @@
+package diskusage
+
+import (
+	"fmt"
+	"testing"
+)
+
+var KB = uint64(1024)
+
+func TestNewDiskUsage(t *testing.T) {
+	usage := NewDiskUsage(".")
+	fmt.Println("Free:", usage.Free()/(KB*KB))
+	fmt.Println("Available:", usage.Available()/(KB*KB))
+	fmt.Println("Size:", usage.Size()/(KB*KB))
+	fmt.Println("Used:", usage.Used()/(KB*KB))
+	fmt.Println("Usage:", usage.Usage()*100, "%")
+}

--- a/src/diskusage/diskusage_windows.go
+++ b/src/diskusage/diskusage_windows.go
@@ -1,0 +1,55 @@
+package diskusage
+
+import (
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+type DiskUsage struct {
+	freeBytes  int64
+	totalBytes int64
+	availBytes int64
+}
+
+// NewDiskUsages returns an object holding the disk usage of volumePath
+// or nil in case of error (invalid path, etc)
+func NewDiskUsage(volumePath string) *DiskUsage {
+	h := windows.MustLoadDLL("kernel32.dll")
+	c := h.MustFindProc("GetDiskFreeSpaceExW")
+
+	du := &DiskUsage{}
+
+	c.Call(
+		uintptr(unsafe.Pointer(windows.StringToUTF16Ptr(volumePath))),
+		uintptr(unsafe.Pointer(&du.freeBytes)),
+		uintptr(unsafe.Pointer(&du.totalBytes)),
+		uintptr(unsafe.Pointer(&du.availBytes)))
+
+	return du
+}
+
+// Free returns total free bytes on file system
+func (du *DiskUsage) Free() uint64 {
+	return uint64(du.freeBytes)
+}
+
+// Available returns total available bytes on file system to an unprivileged user
+func (du *DiskUsage) Available() uint64 {
+	return uint64(du.availBytes)
+}
+
+// Size returns total size of the file system
+func (du *DiskUsage) Size() uint64 {
+	return uint64(du.totalBytes)
+}
+
+// Used returns total bytes used in file system
+func (du *DiskUsage) Used() uint64 {
+	return du.Size() - du.Free()
+}
+
+// Usage returns percentage of use on the file system
+func (du *DiskUsage) Usage() float32 {
+	return float32(du.Used()) / float32(du.Size())
+}

